### PR TITLE
Simplify the Python namespace regular expression

### PR DESCRIPTION
**Pull request by codeinabox**
To address #243

### CODE REVIEW

**Comment by mixxorz** [comment]
Broken on this.

```python
# tests/test_foo.py

class SomeTest(TestCase):

    def test_foo(self):
        foo = date(2017, 11, 16)
```

It resolves as `tests.test_foo.SomeTest.test_foo.date` when run from line 6.

**Comment by codeinabox** [comment]
@mixxorz I've created a test to replicate but I get a different issue

**Comment by michaelmanganiello-eb** [comment]
I'm not familiar with VimScript, but I'll try to be helpful.

I don't get the `'\v^\s*class (Test\w+)|(\w+)\(.*\)'` regex, because the pipe seems to be wrong. It's making `\v^\s*class (Test\w+)` and `(\w+)\(.*\)` two different options. The second one will match to any function call or function definition.

Maybe what we really want is `'\v^\s*class (Test\w+|\w+)\(.*\)'` (look at the difference in the first matching group). However, that could be simplified to `'\v^\s*class (\w+)\(.*\)'`, and, as far as I understand, should make #232 to happen again.

IMHO the real reason for #232 is that we are matching for namespaces even before we already have a match for the test name. For example:

```
1    class TestFoo(BaseTests):
2        def test_foo():
3            class CustomException(Exception):
4                pass
5            assert 1 == 1
```

Should start looking for the namespace *after* the test name was found (line 2).

Of course, we also want to be able to test a whole namespace if we want (e.g. running all the `TestFoo` tests). So I would think the best solution would be to keep this regex as it was before #235 (`'\v^\s*class (\w+)'`), and clean the namespace when we find a test for the first time. That means, setting `namespace` as an empty list here: https://github.com/janko-m/vim-test/blob/612c75eb60948d048122d1f5d0411a59a578a932/autoload/test/base.vim#L82, as we shouldn't have a namespace inside the test. Does that make sense?

**Comment by codeinabox** [comment]
I've managed to get this working for all classes except not matching `class CustomException(Exception)`. @blueyed I know this is one we tried to support but handling this scenario seems tricky without making regular expressions more clever.

File: autoload/test/python.vim [hunk 1]
```diff
@@ -1,4 +1,4 @@
 let test#python#patterns = {
   \ 'test':      ['\v^\s*%(async )?def (test_\w+)'],
-  \ 'namespace': ['\v^\s*class (Test\w+)|(\w+)\(unittest\.TestCase\)'],
+  \ 'namespace': ['\v^\s*class (Test\w+|\w+)'],
```
Comment (janko-m) [inline autoload/test/python.vim]: Note that this regex is equivalent to `'\v^\s*class (\w+)'`.

Comment (codeinabox) [reply autoload/test/python.vim]: Given the issues, do you think we should just simplify the regex down to that and treat classes defined instead a test block as more an edge case?

**Comment by blueyed** [comment]
> clean the namespace when we find a test for the first time
> … as we shouldn't have a namespace inside the test. Does that make sense?

That would solve it, wouldn't it?

**Comment by codeinabox** [comment]
@blueyed yes that would solve it but I'd need to check with @janko-m or review the existing code to find if that's possible.

In the interim it's probably best we simplify the regular expression with this PR so it works for the majority of cases.
